ethexe/contracts/src/Mirror.sol

```diff
@@ -172,14 +176,104 @@ contract Mirror is IMirror {
+
+        if (_message.destination == ETH_EVENT_ADDR && _message.value == 0 && payload.length > 0) {
+            uint256 topicsLength;
+            assembly ("memory-safe") {
+                topicsLength := shr(248, calldataload(payload.offset))
+            }
+
+            if (topicsLength >= 1 && topicsLength <= 4) {
+                uint256 topicsLengthInBytes;
+                unchecked {
+                    topicsLengthInBytes = 1 + topicsLength * 32;
+                }
+
+                if (payload.length >= topicsLengthInBytes) {
+                    bytes32 topic1;
+                    assembly ("memory-safe") {
+                        topic1 := calldataload(add(payload.offset, 1))
+                    }
+
+                    if (
+                        topic1 != StateChanged.selector && topic1 != MessageQueueingRequested.selector
+                            && topic1 != ReplyQueueingRequested.selector && topic1 != ValueClaimingRequested.selector
+                            && topic1 != ExecutableBalanceTopUpRequested.selector && topic1 != Message.selector
+                            && topic1 != Reply.selector && topic1 != ValueClaimed.selector
+                    ) {
+                        uint256 size;
+                        unchecked {
+                            size = payload.length - topicsLengthInBytes;
+                        }
+
+                        uint256 memPtr = Memory.allocate(size);
+                        assembly ("memory-safe") {
+                            calldatacopy(memPtr, add(payload.offset, topicsLengthInBytes), size)
+                        }
+
+                        bytes32 topic2;
+                        bytes32 topic3;
+                        bytes32 topic4;
+                        assembly ("memory-safe") {
+                            topic2 := calldataload(add(payload.offset, 33))
+                            topic3 := calldataload(add(payload.offset, 65))
+                            topic4 := calldataload(add(payload.offset, 97))
+                        }
+
+                        if (topicsLength == 1) {
+                            assembly ("memory-safe") {
+                                log1(memPtr, size, topic1)
+                            }
+                        } else if (topicsLength == 2) {
+                            assembly ("memory-safe") {
+                                log2(memPtr, size, topic1, topic2)
+                            }
+                        } else if (topicsLength == 3) {
+                            assembly ("memory-safe") {
+                                log3(memPtr, size, topic1, topic2, topic3)
+                            }
+                        } else if (topicsLength == 4) {
+                            assembly ("memory-safe") {
+                                log4(memPtr, size, topic1, topic2, topic3, topic4)
+                            }
+                        }
+
+                        return;
```

Comment:
here we also do not emit Message event, which also breaks core protocol logic just to save gas.